service/storage_service.cc

```diff
@@ -2089,6 +2224,78 @@ future<node_ops_cmd_response> storage_service::node_ops_cmd_handler(gms::inet_ad
+            } else if (req.cmd == node_ops_cmd::replace_prepare_mark_alive) {
+                // Wait for local node has marked replacing node as alive
+                size_t nr_replacing_nodes = req.replace_nodes.size();
+                auto start_time = std::chrono::steady_clock::now();
+                auto timeout = std::chrono::seconds(120);
+                for (;;) {
+                    std::vector<gms::inet_address> replacing_nodes_live;
+                    for (auto& x: req.replace_nodes) {
+                        auto replacing_node = x.second;
+                        size_t nr_alive = ss.container().map_reduce0([replacing_node] (auto&& ss) -> size_t {
+                            return ss._gossiper.is_alive(replacing_node) ? 1 : 0;
+                        }, 0, std::plus<size_t>()).get0();
+                        slogger.debug("replace[{}]: Marked replacing_node={} as alive on {} out of {} shards", req.ops_uuid, replacing_node, nr_alive, smp::count);
+                        if (nr_alive == smp::count) {
+                            replacing_nodes_live.push_back(replacing_node);
+                        }
+                    }
+                    slogger.debug("replace[{}]: Waited for marking replacing node as up, replace_nodes={}, replacing_nodes_live={}",
+                            req.ops_uuid, req.replace_nodes, replacing_nodes_live);
+                    if (replacing_nodes_live.size() == nr_replacing_nodes) {
+                        break;
+                    }
+                    if (std::chrono::steady_clock::now() > timeout + start_time) {
+                        throw std::runtime_error(format("Failed to mark replacing node as alive in {} seconds, replace_nodes={}, replacing_nodes_live={}",
+                                timeout.count(), req.replace_nodes, replacing_nodes_live));
+                    }
+                    sleep_abortable(std::chrono::milliseconds(100), ss._abort_source).get();
+                }
+            } else if (req.cmd == node_ops_cmd::replace_prepare_pending_ranges) {
+                // Update the pending_ranges for the replacing node
+                slogger.debug("replace[{}]: Updated pending_ranges from coordinator={}", req.ops_uuid, coordinator);
+                ss.mutate_token_metadata([coordinator, &req, &ss] (mutable_token_metadata_ptr tmptr) mutable {
+                    return ss.update_pending_ranges(tmptr, format("replace {}", req.replace_nodes));
+                }).get();
+            } else if (req.cmd == node_ops_cmd::replace_heartbeat) {
+                slogger.debug("replace[{}]: Updated heartbeat from coordinator={}", req.ops_uuid, coordinator);
+                ss.node_ops_update_heartbeat(ops_uuid);
+            } else if (req.cmd == node_ops_cmd::replace_done) {
+                slogger.info("replace[{}]: Marked ops done from coordinator={}", req.ops_uuid, coordinator);
+                ss.node_ops_done(ops_uuid);
```

Comment:
I've noticed that `node_ops_done` doesn't fail if the op is not found. That may lead to consistency problems where the coordinator thinks the operation succeeded while it failed on some nodes due to watchdog abort.

Consider this scenario:
1. node A is the coordinator for the node op
2. node B is just a regular node 
3. streaming completes
4. node B missed the heartbeat so it rolled back to the previous topology
5. writes were made through B, and they went to the old topology
6. node A sends `replace_done`, which succeeds everywhere, even on node B
7. node A switches the cluster to the new topology via gossip, but writes made in step 5 will not be there (or with smaller CL)

Reply:
In v2, I have already added  extra check about the uuids.

Reply:
What if node B restarts after `replace_done` in step 6? It will revert to the old topology temporarily and run into the same issue, I think

Reply:
If your example, if B missed the heartbeat and rolled back, B will abort the ops with the uuid, delete the host id, the replace_done cmd will FAIL. Node A will faill the ops because it fails at the sening replace_done process. 


If a node restarts after the finish step (node_ops_cmd::replace_done), it is fine because after that, the replacing node will set gossip status to NROMAL. When the node restarts, it will learn the replacing node is a NORMAL node in the cluster.

Reply:
The replacing node sets the gossip status after sending replace_done requests. Node B can restart before it sees the replacing node jumping to NORMAL, because it happens asynchronously.

Reply:
What's wrong with this case? Yes, Node B restarts before it knows the replacing node jumps to NORMAL, after it restarts  will  it learn the NORMAL status from the rest of the cluster. This is exactly what Node B is suppose to see.

Reply:
The problem is that node B, after it restarts, may see and use the old topology for some time window, which can lead to lost writes or inconsistent reads.

Reply:
Please note the old and new topology are the same after replace is done. I do not see how lost of writes or inconsistent reads would happen with v3. If you disagree, please give concrete examples.